b2sdk/utils/curl.py

```diff
@@ -0,0 +1,431 @@
+######################################################################
+#
+# File: b2sdk/utils/curl.py
+#
+# Copyright 2020 Backblaze Inc. All Rights Reserved.
+#
+# License https://www.backblaze.com/using_b2_code.html
+#
+######################################################################
+import collections
+import email.parser
+import threading
+import time
+from dataclasses import dataclass, field
+from functools import partialmethod
+from io import BytesIO
+from typing import Dict, Iterator, List, Optional, Union
+
+import certifi
+import pycurl
+
+from requests.utils import CaseInsensitiveDict
+
+from b2sdk.stream.progress import ReadingStreamWithProgress
+
+
+@dataclass
+class CurlAdapters:
+    adapters: dict = field(default_factory=dict)
+
+    def clear(self):
+        self.adapters = {}
+
+
+@dataclass
+class CurlRequest:
+    url: str
+    headers: CaseInsensitiveDict
+
+
+class StreamedBytes:
+    """
+    Structure that works "kinda" like an unlimited ring buffer.
+
+    We can put the data on one side and read it from the other,
+    and the data read is lost (memory is released).
+    """
+
+    def __init__(self):
+        # We store our data in chunks in a deque. Whenever data is requested,
+        # we remove them from deque up to the size, and put back the reminder.
+        self.deque = collections.deque()
+        self.lock = threading.Lock()
+        self.first_write_event = threading.Event()
+
+    def is_empty(self) -> bool:
+        with self.lock:
+            return len(self.deque) == 0
+
+    def write(self, buffer) -> int:
+        with self.lock:
+            if len(buffer) == 0:
+                return 0
+
+            # Whole buffer is stored as deque is implemented as double-linked list.
+            # This way we have fewer, (hopefully) larger elements.
+            self.deque.append(buffer)
+
+            # If we haven't informed everyone yet, it's the time. We've received
+            # first buffer of data. This is used to indicate that all the headers
+            # are already downloaded.
+            if not self.first_write_event.is_set():
```

Comment:
wait, so HEAD requests are not supported by this?

Reply:
`HEAD` should never contain content, so no. `HEAD` requests in streamed mode are just informed by `CurlManager` that they are "done", and then they fetch headers and status code.